Index: community/platform/core-api/src/com/intellij/openapi/application/ApplicationNamesInfo.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- community/platform/core-api/src/com/intellij/openapi/application/ApplicationNamesInfo.java	(revision d813397736c0afaa381111c73e724e6accb0006d)
+++ community/platform/core-api/src/com/intellij/openapi/application/ApplicationNamesInfo.java	(date 1651654462417)
@@ -68,7 +68,7 @@
       }
 
       // production
-      String appInfoData = getAppInfoData();
+      String appInfoData = "";
       if (!appInfoData.isEmpty()) {
         return XmlDomReader.readXmlAsModel(appInfoData.getBytes(StandardCharsets.UTF_8));
       }
