Index: community/platform/core-api/src/com/intellij/openapi/project/Project.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- community/platform/core-api/src/com/intellij/openapi/project/Project.java	(revision 83444e03e40e6673d2650a7840140cca65d5b5f6)
+++ community/platform/core-api/src/com/intellij/openapi/project/Project.java	(revision )
@@ -32,7 +32,7 @@
  * <p>To get the list of all open projects, use {@code ProjectManager.getInstance().getOpenProjects()}.
  */
 public interface Project extends ComponentManager, AreaInstance {
-  String DIRECTORY_STORE_FOLDER = ".idea";
+  String DIRECTORY_STORE_FOLDER = ".mps";
 
   /**
    * Returns a name ot the project. For a directory-based project it's an arbitrary string specified by user at project creation

Index: platform/configuration-store-impl/src/ProjectStoreImpl.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/platform/configuration-store-impl/src/ProjectStoreImpl.kt b/platform/configuration-store-impl/src/ProjectStoreImpl.kt
--- a/platform/configuration-store-impl/src/ProjectStoreImpl.kt	(revision 5fae0bafa59db4d0adedcbfaeeb394647dce2959)
+++ b/platform/configuration-store-impl/src/ProjectStoreImpl.kt	(date 1708600813429)
@@ -93,11 +93,11 @@
     val storageManager = storageManager
     val isUnitTestMode = ApplicationManager.getApplication().isUnitTestMode
     val macros = ArrayList<Macro>(5)
-    if (file.toString().endsWith(ProjectFileType.DOT_DEFAULT_EXTENSION)) {
+    if (file.toString().endsWith(".mpr"/*ProjectFileType.DOT_DEFAULT_EXTENSION*/)) {
       macros.add(Macro(PROJECT_FILE, file))
 
       val workspacePath = file.parent.resolve("${file.fileName.toString()
-        .removeSuffix(ProjectFileType.DOT_DEFAULT_EXTENSION)}${WorkspaceFileType.DOT_DEFAULT_EXTENSION}")
+        .removeSuffix(".mpr"/*ProjectFileType.DOT_DEFAULT_EXTENSION*/)}${WorkspaceFileType.DOT_DEFAULT_EXTENSION}")
       macros.add(Macro(StoragePathMacros.WORKSPACE_FILE, workspacePath))
 
       if (isUnitTestMode) {

Index: platform/ide-core/src/com/intellij/openapi/project/ProjectUtil.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/platform/ide-core/src/com/intellij/openapi/project/ProjectUtil.kt b/platform/ide-core/src/com/intellij/openapi/project/ProjectUtil.kt
--- a/platform/ide-core/src/com/intellij/openapi/project/ProjectUtil.kt	(revision 858b2cd664da5ab63421443a2b30f3a73da40ca7)
+++ b/platform/ide-core/src/com/intellij/openapi/project/ProjectUtil.kt	(date 1704800610080)
@@ -202,7 +202,7 @@
     else -> {
       // the lower case here is used for cosmetic reasons (develar - discussed with jeka - leave it as it was,
       // user projects will not have long names as in our tests
-      PathUtilRt.getFileName(presentableUrl).lowercase(Locale.US).removeSuffix(ProjectFileType.DOT_DEFAULT_EXTENSION)
+      PathUtilRt.getFileName(presentableUrl).lowercase(Locale.US).removeSuffix(".mpr"/*ProjectFileType.DOT_DEFAULT_EXTENSION*/)
     }
   }
   name = sanitizeFileName(name, truncateIfNeeded = false)

Index: jps/model-serialization/src/org/jetbrains/jps/model/serialization/PathMacroUtil.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/jps/model-serialization/src/org/jetbrains/jps/model/serialization/PathMacroUtil.java b/jps/model-serialization/src/org/jetbrains/jps/model/serialization/PathMacroUtil.java
--- a/jps/model-serialization/src/org/jetbrains/jps/model/serialization/PathMacroUtil.java	(revision fce47f00ab09add5f5df856501530672b9e7aa95)
+++ b/jps/model-serialization/src/org/jetbrains/jps/model/serialization/PathMacroUtil.java	(date 1726584055744)
@@ -31,7 +31,7 @@
   public static final String MODULE_WORKING_DIR = "$" + MODULE_WORKING_DIR_NAME + "$";

   @ApiStatus.Internal
-  @NonNls public static final String DIRECTORY_STORE_NAME = ".idea";
+  @NonNls public static final String DIRECTORY_STORE_NAME = ".mps";
   @ApiStatus.Internal
   @NonNls public static final String APPLICATION_HOME_DIR = "APPLICATION_HOME_DIR";
   @ApiStatus.Internal
