Index: platform/execution-impl/src/com/intellij/execution/actions/RunConfigurationsComboBoxAction.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/platform/execution-impl/src/com/intellij/execution/actions/RunConfigurationsComboBoxAction.java b/platform/execution-impl/src/com/intellij/execution/actions/RunConfigurationsComboBoxAction.java
--- a/platform/execution-impl/src/com/intellij/execution/actions/RunConfigurationsComboBoxAction.java	(revision 89fc0bfc19af35dc228bbc7d07dd247a069a85a1)
+++ b/platform/execution-impl/src/com/intellij/execution/actions/RunConfigurationsComboBoxAction.java	(date 1688119691002)
@@ -60,6 +60,7 @@
       return false;
     }
 
+    if (true) return false;
     if (PlatformUtils.isIntelliJ()) return true;
     if (PlatformUtils.isPhpStorm()) return true;
     if (PlatformUtils.isWebStorm()) return true;
