Index: platform/util/ui/src/com/intellij/openapi/util/IconLoader.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/platform/util/ui/src/com/intellij/openapi/util/IconLoader.kt b/platform/util/ui/src/com/intellij/openapi/util/IconLoader.kt
--- a/platform/util/ui/src/com/intellij/openapi/util/IconLoader.kt	(revision fce47f00ab09add5f5df856501530672b9e7aa95)
+++ b/platform/util/ui/src/com/intellij/openapi/util/IconLoader.kt	(date 1726580912566)
@@ -318,6 +318,7 @@
     return replacer.replaceIcon(icon)
   }
 
+  @JvmStatic
   fun detachClassLoader(classLoader: ClassLoader) {
     iconCache.asMap().entries.removeIf { (key, icon) ->
       icon.detachClassLoader(classLoader) || key.second === classLoader
 