Index: community/platform/platform-impl/src/com/intellij/idea/StartupUtil.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- community/platform/platform-impl/src/com/intellij/idea/StartupUtil.java	(revision 9d26416bebfdc0b27bc49afdacf204d5ff9d2ac7)
+++ community/platform/platform-impl/src/com/intellij/idea/StartupUtil.java	(revision ea002acba0d46b34790d2ca7d4d7f3e6386295a7)
@@ -109,15 +109,21 @@
     return serverFuture == null ? CompletableFuture.completedFuture(null) : serverFuture;
   }
 
-  private static @Nullable Object loadEuaDocument() {
-    if (!ApplicationInfoImpl.getShadowInstance().isVendorJetBrains()) {
-      return null;
-    }
+  private static @NotNull Future<@Nullable Object> loadEuaDocument(@NotNull ExecutorService executorService) {
+    if (Main.isHeadless()) {
+      return CompletableFuture.completedFuture(null);
+    }
+
+    return executorService.submit(() -> {
+      if (!ApplicationInfoImpl.getShadowInstance().isVendorJetBrains()) {
+        return null;
+      }
 
-    Activity euaActivity = StartUpMeasurer.startActivity("eua getting");
-    EndUserAgreement.Document result = EndUserAgreement.getLatestDocument();
-    euaActivity.end();
-    return result;
+      Activity euaActivity = StartUpMeasurer.startActivity("eua getting");
+      EndUserAgreement.Document result = EndUserAgreement.getLatestDocument();
+      euaActivity.end();
+      return result;
+    });
   }
 
   public interface AppStarter {
@@ -178,7 +184,7 @@
 
     activity = activity.endAndStart("LaF init scheduling");
     // EndUserAgreement.Document type is not specified to avoid class loading
-    Future<Object> euaDocument = Main.isHeadless() ? null : executorService.submit(StartupUtil::loadEuaDocument);
+    Future<Object> euaDocument = loadEuaDocument(executorService);
     CompletableFuture<?> initUiTask = scheduleInitUi(args, executorService, euaDocument);
     activity.end();
 
@@ -245,10 +251,11 @@
                                @NotNull Logger log,
                                boolean configImportNeeded,
                                @NotNull Future<AppStarter> appStarterFuture,
-                               @Nullable Future<Object> euaDocument) throws Exception {
+                               @NotNull Future<@Nullable Object> euaDocument) throws Exception {
     if (!Main.isHeadless()) {
       Activity activity = StartUpMeasurer.startMainActivity("eua showing");
-      boolean agreementDialogWasShown = euaDocument != null && showUserAgreementAndConsentsIfNeeded(log, initUiTask, euaDocument);
+      Object document = euaDocument.get();
+      boolean agreementDialogWasShown = document != null && showUserAgreementAndConsentsIfNeeded(log, initUiTask, (EndUserAgreement.Document)document);
 
       if (agreementDialogWasShown) {
         activity = activity.endAndStart("screen reader checking");
@@ -293,7 +300,7 @@
     }
   }
 
-  private static @NotNull CompletableFuture<?> scheduleInitUi(@NotNull String @NotNull [] args, @NotNull ExecutorService executor, @Nullable Future<Object> eulaDocument) {
+  private static @NotNull CompletableFuture<?> scheduleInitUi(@NotNull String @NotNull [] args, @NotNull Executor executor, @NotNull Future<@Nullable Object> eulaDocument) {
     // mainly call sun.util.logging.PlatformLogger.getLogger - it takes enormous time (up to 500 ms)
     // Before lockDirsAndConfigureLogger can be executed only tasks that do not require log,
     // because we don't want to complicate logging. It is OK, because lockDirsAndConfigureLogger is not so heavy-weight as UI tasks.
@@ -338,7 +345,8 @@
               Activity eulaActivity = prepareSplashActivity.startChild("splash eula isAccepted");
               boolean isEulaAccepted;
               try {
-                isEulaAccepted = eulaDocument == null || ((EndUserAgreement.Document)eulaDocument.get()).isAccepted();
+                EndUserAgreement.Document document = (EndUserAgreement.Document)eulaDocument.get();
+                isEulaAccepted = document == null || document.isAccepted();
               }
               catch (InterruptedException | ExecutionException ignore) {
                 isEulaAccepted = true;
@@ -531,7 +539,7 @@
     }
   }
 
-  private static void lockSystemDirs(@NotNull Path configPath, @NotNull Path systemPath, @NotNull String[] args) throws Exception {
+  private static void lockSystemDirs(@NotNull Path configPath, @NotNull Path systemPath, @NotNull String @NotNull[] args) throws Exception {
     if (ourSocketLock != null) {
       throw new AssertionError("Already initialized");
     }
@@ -573,6 +581,7 @@
       Logger.setFactory(new LoggerFactory());
     }
     catch (Exception e) {
+      //noinspection CallToPrintStackTrace
       e.printStackTrace();
     }
     Logger log = Logger.getInstance(Main.class);
@@ -757,10 +766,9 @@
 
   private static boolean showUserAgreementAndConsentsIfNeeded(@NotNull Logger log,
                                                               @NotNull CompletableFuture<?> initUiTask,
-                                                              @NotNull Future<Object> euaDocument) throws ExecutionException, InterruptedException {
+                                                              @NotNull EndUserAgreement.Document agreement) {
     boolean dialogWasShown = false;
     EndUserAgreement.updateCachedContentToLatestBundledVersion();
-    EndUserAgreement.Document agreement = (EndUserAgreement.Document)euaDocument.get();
     if (!agreement.isAccepted()) {
       // todo: does not seem to request focus when shown
       runInEdtAndWait(log, () -> Agreements.INSTANCE.showEndUserAndDataSharingAgreements(agreement), initUiTask);
