Index: community/platform/platform-impl/src/com/intellij/idea/StartupUtil.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- community/platform/platform-impl/src/com/intellij/idea/StartupUtil.java	(revision 68cbe80d4547b1bf069bd2a66885bb3b44837984)
+++ community/platform/platform-impl/src/com/intellij/idea/StartupUtil.java	(revision e114724d084412268b8156b31b04c3d8f9175aef)
@@ -106,15 +106,21 @@
     return serverFuture == null ? CompletableFuture.completedFuture(null) : serverFuture;
   }
 
-  private static @Nullable Object loadEuaDocument() {
-    if (!ApplicationInfoImpl.getShadowInstance().isVendorJetBrains()) {
-      return null;
-    }
+  private static @NotNull Future<@Nullable Object> loadEuaDocument(@NotNull ExecutorService executorService) {
+    if (Main.isHeadless()) {
+      return CompletableFuture.completedFuture(null);
+    }
+
+    return executorService.submit(() -> {
+      if (!ApplicationInfoImpl.getShadowInstance().isVendorJetBrains()) {
+        return null;
+      }
 
-    Activity euaActivity = StartUpMeasurer.startActivity("eua getting");
-    EndUserAgreement.Document result = EndUserAgreement.getLatestDocument();
-    euaActivity.end();
-    return result;
+      Activity euaActivity = StartUpMeasurer.startActivity("eua getting");
+      EndUserAgreement.Document result = EndUserAgreement.getLatestDocument();
+      euaActivity.end();
+      return result;
+    });
   }
 
   public interface AppStarter {
@@ -175,7 +181,7 @@
 
     activity = activity.endAndStart("LaF init scheduling");
     // EndUserAgreement.Document type is not specified to avoid class loading
-    Future<Object> euaDocument = Main.isHeadless() ? null : executorService.submit(StartupUtil::loadEuaDocument);
+    Future<Object> euaDocument = loadEuaDocument(executorService);
     CompletableFuture<?> initUiTask = scheduleInitUi(args, executorService, euaDocument);
     activity.end();
 
@@ -248,10 +254,11 @@
                                @NotNull Logger log,
                                boolean configImportNeeded,
                                @NotNull Future<AppStarter> appStarterFuture,
-                               @Nullable Future<Object> euaDocument) throws Exception {
+                               @NotNull Future<@Nullable Object> euaDocument) throws Exception {
     if (!Main.isHeadless()) {
       Activity activity = StartUpMeasurer.startMainActivity("eua showing");
-      boolean agreementDialogWasShown = euaDocument != null && showUserAgreementAndConsentsIfNeeded(log, initUiTask, euaDocument);
+      Object document = euaDocument.get();
+      boolean agreementDialogWasShown = document != null && showUserAgreementAndConsentsIfNeeded(log, initUiTask, (EndUserAgreement.Document)document);
       if (configImportNeeded) {
         activity = activity.endAndStart("config importing");
         AppStarter appStarter = getAppStarter(appStarterFuture);
@@ -293,8 +300,7 @@
     }
   }
 
-  @NotNull
-  private static CompletableFuture<?> scheduleInitUi(@NotNull String @NotNull [] args, @NotNull ExecutorService executor, @Nullable Future<Object> eulaDocument) {
+  private static @NotNull CompletableFuture<?> scheduleInitUi(@NotNull String @NotNull [] args, @NotNull Executor executor, @NotNull Future<@Nullable Object> eulaDocument) {
     // mainly call sun.util.logging.PlatformLogger.getLogger - it takes enormous time (up to 500 ms)
     // Before lockDirsAndConfigureLogger can be executed only tasks that do not require log,
     // because we don't want to complicate logging. It is OK, because lockDirsAndConfigureLogger is not so heavy-weight as UI tasks.
@@ -339,7 +345,8 @@
               Activity eulaActivity = prepareSplashActivity.startChild("splash eula isAccepted");
               boolean isEulaAccepted;
               try {
-                isEulaAccepted = eulaDocument == null || ((EndUserAgreement.Document)eulaDocument.get()).isAccepted();
+                EndUserAgreement.Document document = (EndUserAgreement.Document)eulaDocument.get();
+                isEulaAccepted = document == null || document.isAccepted();
               }
               catch (InterruptedException | ExecutionException ignore) {
                 isEulaAccepted = true;
@@ -532,7 +539,7 @@
     }
   }
 
-  private static synchronized void lockSystemDirs(String[] args) throws Exception {
+  private static synchronized void lockSystemDirs(String @NotNull[] args) throws Exception {
     if (ourSocketLock != null) throw new AssertionError("Already initialized");
     ourSocketLock = new SocketLock(PathManager.getConfigPath(), PathManager.getSystemPath());
 
@@ -567,9 +574,14 @@
   }
 
   @SuppressWarnings("UseOfSystemOutOrSystemErr")
-  @NotNull
-  private static Logger setupLogger() {
-    Logger.setFactory(new LoggerFactory());
+  private static @NotNull Logger setupLogger() {
+    try {
+      Logger.setFactory(new LoggerFactory());
+    }
+    catch (Exception e) {
+      //noinspection CallToPrintStackTrace
+      e.printStackTrace();
+    }
     Logger log = Logger.getInstance(Main.class);
     log.info("------------------------------------------------------ IDE STARTED ------------------------------------------------------");
     ShutDownTracker.getInstance().registerShutdownTask(() -> {
@@ -731,10 +743,9 @@
 
   private static boolean showUserAgreementAndConsentsIfNeeded(@NotNull Logger log,
                                                               @NotNull CompletableFuture<?> initUiTask,
-                                                              @NotNull Future<Object> euaDocument) throws ExecutionException, InterruptedException {
+                                                              @NotNull EndUserAgreement.Document agreement) {
     boolean dialogWasShown = false;
     EndUserAgreement.updateCachedContentToLatestBundledVersion();
-    EndUserAgreement.Document agreement = (EndUserAgreement.Document)euaDocument.get();
     if (!agreement.isAccepted()) {
       // todo: does not seem to request focus when shown
       runInEdtAndWait(log, () -> Agreements.INSTANCE.showEndUserAndDataSharingAgreements(agreement), initUiTask);
