Index: platform/platform-impl/src/com/intellij/openapi/wm/impl/ProjectFrameHelper.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/platform/platform-impl/src/com/intellij/openapi/wm/impl/ProjectFrameHelper.kt b/platform/platform-impl/src/com/intellij/openapi/wm/impl/ProjectFrameHelper.kt
--- a/platform/platform-impl/src/com/intellij/openapi/wm/impl/ProjectFrameHelper.kt	(revision 7ff5c6541b265b30d4ecd702dccd8005502d3a74)
+++ b/platform/platform-impl/src/com/intellij/openapi/wm/impl/ProjectFrameHelper.kt	(date 1679411795820)
@@ -199,14 +199,17 @@
     if (SystemInfoRt.isMac) {
       frame.iconImage = null
     }
-    else if (SystemInfoRt.isLinux) {
-      frame.addComponentListener(object : ComponentAdapter() {
-        override fun componentShown(e: ComponentEvent) {
-          frame.removeComponentListener(this)
-          IdeMenuBar.installAppMenuIfNeeded(frame)
-        }
-      })
-      // in production (not from sources) makes sense only on Linux
+    else {
+      if (SystemInfoRt.isLinux) {
+        frame.addComponentListener(object : ComponentAdapter() {
+          override fun componentShown(e: ComponentEvent) {
+            frame.removeComponentListener(this)
+            IdeMenuBar.installAppMenuIfNeeded(frame)
+          }
+        })
+      }
+      // in production (not from sources) it makes sense only on Linux
+      // or on Windows (for products that don't use a native launcher, e.g. MPS)
       AppUIUtil.updateWindowIcon(frame)
     }
     MouseGestureManager.getInstance().add(this)
