Index: platform/util/ui/src/com/intellij/ui/scale/JBUIScale.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/platform/util/ui/src/com/intellij/ui/scale/JBUIScale.kt b/platform/util/ui/src/com/intellij/ui/scale/JBUIScale.kt
--- a/platform/util/ui/src/com/intellij/ui/scale/JBUIScale.kt	(revision f34dfce5ba37a9cc56515a078ba587476ae2fd20)
+++ b/platform/util/ui/src/com/intellij/ui/scale/JBUIScale.kt	(date 1686640605189)
@@ -105,6 +105,10 @@
     }
 
     val log = thisLogger()
+    if (font == null && SystemInfoRt.isLinux) {
+      font = Font(Font.SANS_SERIF, Font.PLAIN, 15);
+      log.warn("Font was not loaded by Label.font key, fall back to SansSerif");
+    }
     val isScaleVerbose = SCALE_VERBOSE
     if (isScaleVerbose) {
       log.info("Label font: ${font.fontName}, ${font.size}")
