Index: platform/platform-impl/src/com/intellij/ide/gdpr/EndUserAgreement.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/platform/platform-impl/src/com/intellij/ide/gdpr/EndUserAgreement.java b/platform/platform-impl/src/com/intellij/ide/gdpr/EndUserAgreement.java
--- a/platform/platform-impl/src/com/intellij/ide/gdpr/EndUserAgreement.java	(revision 5745d2a0337ff4fe9221b43fb3ade833feeeee81)
+++ b/platform/platform-impl/src/com/intellij/ide/gdpr/EndUserAgreement.java	(date 1756349521721)
@@ -29,6 +29,8 @@
   private static final String CWM_GUEST_EULA_NAME = "cwmGuestEua";
   private static final String EULA_DOCUMENT_NAME = "eua";
   private static final String EULA_COMMUNITY_DOCUMENT_NAME = "euaCommunity";
+  private static final String EULA_MPS_DOCUMENT_NAME = "euaMps";
+  private static final String EULA_MPS_CUSTOM_DOCUMENT_KEY = "mps.eua.document.name";
   private static final String EULA_EAP_DOCUMENT_NAME = EULA_DOCUMENT_NAME + "Eap";

   private static final String PRIVACY_POLICY_CONTENT_FILE_NAME = "Cached";
@@ -201,6 +203,12 @@
       if (PlatformUtils.isCommunityEdition()) {
         return shouldUseEAPAgreement()? DEFAULT_DOC_EAP_NAME : EULA_COMMUNITY_DOCUMENT_NAME;
       }
+      if (PlatformUtils.isMPS()) {
+        if (isValidFileName(System.getProperty(EULA_MPS_CUSTOM_DOCUMENT_KEY))) {
+          return System.getProperty(EULA_MPS_CUSTOM_DOCUMENT_KEY);
+        }
+        return shouldUseEAPAgreement()? DEFAULT_DOC_EAP_NAME : EULA_MPS_DOCUMENT_NAME;
+      }
       if (PlatformUtils.isJetBrainsClient()) {
         return CWM_GUEST_EULA_NAME;
       }
