Index: platform/testFramework/common/src/common/testApplication.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/platform/testFramework/common/src/common/testApplication.kt b/platform/testFramework/common/src/common/testApplication.kt
--- a/platform/testFramework/common/src/common/testApplication.kt	(revision 14aa4be4a15c6eec42d2c7a089c82be24313e295)
+++ b/platform/testFramework/common/src/common/testApplication.kt	(revision f85c84e4d7e2f0f25eddda91a06a4d74df437e90)
@@ -177,7 +177,8 @@
 
   try {
     // 40 seconds - tests maybe executed on cloud agents where I/O is very slow
-    val pluginSet = loadedModuleFuture.asCompletableFuture().get(40, TimeUnit.SECONDS)
+    val timeout = System.getProperty("intellij.testFramework.modules.timeout.seconds", "40").toLong()
+    val pluginSet = loadedModuleFuture.asCompletableFuture().get(timeout, TimeUnit.SECONDS)
     app.registerComponents(modules = pluginSet.getEnabledModules(), app = app)
 
     val task = suspend {
@@ -210,7 +211,8 @@
 
 private suspend fun preloadServicesAndCallAppInitializedListeners(app: ApplicationImpl) {
   coroutineScope {
-    withTimeout(Duration.ofSeconds(40).toMillis()) {
+    val timeout = System.getProperty("intellij.testFramework.services.timeout.seconds", "40").toLong()
+    withTimeout(Duration.ofSeconds(timeout).toMillis()) {
       val pathMacroJob = preloadCriticalServices(
         app = app,
         asyncScope = app.getCoroutineScope(),
 