Index: platform/platform-impl/src/com/intellij/openapi/actionSystem/impl/ActionToolbarImpl.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/platform/platform-impl/src/com/intellij/openapi/actionSystem/impl/ActionToolbarImpl.java b/platform/platform-impl/src/com/intellij/openapi/actionSystem/impl/ActionToolbarImpl.java
--- a/platform/platform-impl/src/com/intellij/openapi/actionSystem/impl/ActionToolbarImpl.java	(revision 6dba988aa708f03fc5fb0fd6ab695556295deaf8)
+++ b/platform/platform-impl/src/com/intellij/openapi/actionSystem/impl/ActionToolbarImpl.java	(date 1668598293007)
@@ -498,7 +498,7 @@
   }
 
   protected @NotNull ActionButton createToolbarButton(@NotNull AnAction action,
-                                                      final ActionButtonLook look,
+                                                      @Nullable ActionButtonLook look,
                                                       @NotNull String place,
                                                       @NotNull Presentation presentation,
                                                       @NotNull Dimension minimumSize) {
Index: platform/platform-impl/src/com/intellij/openapi/wm/impl/welcomeScreen/ProjectsTabFactory.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/platform/platform-impl/src/com/intellij/openapi/wm/impl/welcomeScreen/ProjectsTabFactory.kt b/platform/platform-impl/src/com/intellij/openapi/wm/impl/welcomeScreen/ProjectsTabFactory.kt
--- a/platform/platform-impl/src/com/intellij/openapi/wm/impl/welcomeScreen/ProjectsTabFactory.kt	(revision 6dba988aa708f03fc5fb0fd6ab695556295deaf8)
+++ b/platform/platform-impl/src/com/intellij/openapi/wm/impl/welcomeScreen/ProjectsTabFactory.kt	(date 1668598293022)
@@ -201,7 +201,7 @@
     toolbarActionGroup.addAction(moreActionGroup)
     val toolbar: ActionToolbarImpl = object : ActionToolbarImpl(ActionPlaces.WELCOME_SCREEN, toolbarActionGroup, true) {
       override fun createToolbarButton(action: AnAction,
-                                       look: ActionButtonLook,
+                                       look: ActionButtonLook?,
                                        place: String,
                                        presentation: Presentation,
                                        minimumSize: Dimension): ActionButton {
