Index: platform/platform-impl/src/com/intellij/ide/impl/ProjectUtil.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- community/platform/platform-impl/src/com/intellij/ide/impl/ProjectUtil.java	(revision 9900ca50c9451db9b00ba7536220393b1345df53)
+++ community/platform/platform-impl/src/com/intellij/ide/impl/ProjectUtil.java	(date 1623669395948)
@@ -665,7 +665,7 @@
   private static String getUserHomeProjectDir() {
     final String userHome = SystemProperties.getUserHome();
     String productName = ApplicationNamesInfo.getInstance().getLowercaseProductName();
-    if (PlatformUtils.isCLion() || PlatformUtils.isAppCode() || PlatformUtils.isDataGrip()) {
+    if (PlatformUtils.isCLion() || PlatformUtils.isAppCode() || PlatformUtils.isDataGrip() || "MPS".equals(ApplicationNamesInfo.getInstance().getProductName())) {
       productName = ApplicationNamesInfo.getInstance().getProductName();
     }
     return userHome.replace('/', File.separatorChar) + File.separator + productName + "Projects";
