Index: platform/platform-impl/src/com/intellij/ide/impl/ProjectUtil.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/platform/platform-impl/src/com/intellij/ide/impl/ProjectUtil.kt b/platform/platform-impl/src/com/intellij/ide/impl/ProjectUtil.kt
--- a/platform/platform-impl/src/com/intellij/ide/impl/ProjectUtil.kt	(revision 1d27236ae89e91e502e388d0c11c912f51e152a3)
+++ b/platform/platform-impl/src/com/intellij/ide/impl/ProjectUtil.kt	(date 1665567129435)
@@ -495,7 +495,7 @@
 
   @JvmStatic
   fun getUserHomeProjectDir(): String {
-    val productName = if (PlatformUtils.isCLion() || PlatformUtils.isAppCode() || PlatformUtils.isDataGrip()) {
+    val productName = if (PlatformUtils.isCLion() || PlatformUtils.isAppCode() || PlatformUtils.isDataGrip() || "MPS" == ApplicationNamesInfo.getInstance().productName) {
       ApplicationNamesInfo.getInstance().productName
     }
     else {
