Index: platform/platform-impl/src/com/intellij/ide/impl/ProjectUtil.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- community/platform/platform-impl/src/com/intellij/ide/impl/ProjectUtil.java	(revision c29a202eba261ddf5e9872ecfabfb09bd53b0136)
+++ community/platform/platform-impl/src/com/intellij/ide/impl/ProjectUtil.java	(date 1634821425810)
@@ -639,7 +639,8 @@

   private static String getUserHomeProjectDir() {
     String productName;
-    if (PlatformUtils.isCLion() || PlatformUtils.isAppCode() || PlatformUtils.isDataGrip()) {
+    if (PlatformUtils.isCLion() || PlatformUtils.isAppCode() || PlatformUtils.isDataGrip() ||
+        "MPS".equals(ApplicationNamesInfo.getInstance().getProductName())) {
       productName = ApplicationNamesInfo.getInstance().getProductName();
     }
     else {
