Index: community/platform/core-impl/src/com/intellij/ide/plugins/PluginManagerCore.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- community/platform/core-impl/src/com/intellij/ide/plugins/PluginManagerCore.java	(revision 6fb60ed87003a79708d94d963ed3f339e0382111)
+++ community/platform/core-impl/src/com/intellij/ide/plugins/PluginManagerCore.java	(revision d38cce846cab199bf2251ac4acfe0781eff4d8b4)
@@ -1220,6 +1220,7 @@
     Set<PluginId> explicitlyEnabled = null;
     if (selectedIds != null) {
       HashSet<String> set = new HashSet<>(StringUtil.split(selectedIds, ","));
+      set.addAll(((ApplicationInfoImpl)ApplicationInfoImpl.getShadowInstance()).getEssentialPluginsIds());
       explicitlyEnabled = JBIterable.from(allDescriptors)
         .map(IdeaPluginDescriptor::getPluginId)
         .filter(o -> set.contains(o.getIdString())).addAllTo(new LinkedHashSet<>());
@@ -1245,9 +1246,11 @@
       }
       else if (explicitlyEnabled != null) {
         if (!explicitlyEnabled.contains(descriptor.getPluginId())) {
-          errorSuffix = selectedIds != null
-                        ? "is not in 'idea.load.plugins.id' system property"
-                        : "category doesn't match 'idea.load.plugins.category' system property";
+          errorSuffix = "";
+          getLogger().info("Plugin " + toPresentableName(descriptor) + " " +
+                           (selectedIds != null
+                            ? "is not in 'idea.load.plugins.id' system property"
+                            : "category doesn't match 'idea.load.plugins.category' system property"));
         }
         else {
           errorSuffix = null;
