Index: platform/core-impl/src/com/intellij/ide/plugins/PluginManagerCore.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/platform/core-impl/src/com/intellij/ide/plugins/PluginManagerCore.kt b/platform/core-impl/src/com/intellij/ide/plugins/PluginManagerCore.kt
--- a/platform/core-impl/src/com/intellij/ide/plugins/PluginManagerCore.kt	(revision 4b67311828f3673ec1b97788764bc1f8ea527253)
+++ b/platform/core-impl/src/com/intellij/ide/plugins/PluginManagerCore.kt	(date 1697452886328)
@@ -436,7 +436,7 @@
         val descriptor = idMap[id] ?: continue
         selectedPlugins.add(descriptor)
         processAllNonOptionalDependencies(rootDescriptor = descriptor, pluginIdMap = idMap, consumer = { dependency ->
-          selectedPlugins.add(dependency!!)
+          if (dependency != null) selectedPlugins.add(dependency)
           FileVisitResult.CONTINUE
         })
       }
