Index: platform/platform-impl/src/com/intellij/ide/impl/TrustedPathsSettings.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/platform/platform-impl/src/com/intellij/ide/impl/TrustedPathsSettings.kt b/platform/platform-impl/src/com/intellij/ide/impl/TrustedPathsSettings.kt
--- a/platform/platform-impl/src/com/intellij/ide/impl/TrustedPathsSettings.kt	(revision 6dba988aa708f03fc5fb0fd6ab695556295deaf8)
+++ b/platform/platform-impl/src/com/intellij/ide/impl/TrustedPathsSettings.kt	(date 1665567129440)
@@ -12,9 +12,10 @@
 @ApiStatus.Internal
 fun isPathTrustedInSettings(path: Path): Boolean = service<TrustedPathsSettings>().isPathTrusted(path)
 
+@ApiStatus.Internal
 @State(name = "Trusted.Paths.Settings", storages = [Storage(value = "trusted-paths.xml", roamingType = RoamingType.DISABLED)])
 @Service(Service.Level.APP)
-internal class TrustedPathsSettings : SimplePersistentStateComponent<TrustedPathsSettings.State>(State()) {
+class TrustedPathsSettings : SimplePersistentStateComponent<TrustedPathsSettings.State>(State()) {
   class State : BaseState() {
     @get:OptionTag("TRUSTED_PATHS")
     var trustedPaths by list<String>()
