Index: community/platform/build-scripts/groovy/org/jetbrains/intellij/build/impl/DistributionJARsBuilder.groovy
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- community/platform/build-scripts/groovy/org/jetbrains/intellij/build/impl/DistributionJARsBuilder.groovy	(revision f3337ede11680b745f513943772150ee781a51c0)
+++ community/platform/build-scripts/groovy/org/jetbrains/intellij/build/impl/DistributionJARsBuilder.groovy	(date 1633628534316)
@@ -454,7 +454,7 @@
   }
 
   void processLibDirectoryLayout(LayoutBuilder layoutBuilder, ProjectStructureMapping projectStructureMapping, boolean copyFiles) {
-    if (copyFiles) {
+    if (false) {
       Path moduleOutDir = Path.of(buildContext.getModuleOutputPath(buildContext.findRequiredModule("intellij.platform.core")))
       Path patchedClassFileRoot = buildContext.paths.tempDir.resolve("appInfoData")
       Path classRelativeFile = Path.of("com/intellij/openapi/application/ApplicationNamesInfo.class")
