Index: community/platform/platform-impl/src/com/intellij/openapi/fileEditor/impl/text/FileDropHandler.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- community/platform/platform-impl/src/com/intellij/openapi/fileEditor/impl/text/FileDropHandler.java	(date 1461158849000)
+++ community/platform/platform-impl/src/com/intellij/openapi/fileEditor/impl/text/FileDropHandler.java	(revision )
@@ -16,9 +16,11 @@
 package com.intellij.openapi.fileEditor.impl.text;

 import com.intellij.ide.dnd.FileCopyPasteUtil;
+import com.intellij.openapi.editor.CustomFileDropHandler;
 import com.intellij.openapi.editor.Document;
 import com.intellij.openapi.editor.Editor;
 import com.intellij.openapi.editor.EditorDropHandler;
+import com.intellij.openapi.extensions.Extensions;
 import com.intellij.openapi.fileEditor.*;
 import com.intellij.openapi.fileEditor.ex.FileEditorManagerEx;
 import com.intellij.openapi.fileEditor.impl.EditorWindow;
@@ -26,6 +28,7 @@
 import com.intellij.openapi.project.Project;
 import com.intellij.openapi.vfs.LocalFileSystem;
 import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.util.containers.ContainerUtil;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;

@@ -54,6 +57,11 @@
     if (project != null) {
       final List<File> fileList = FileCopyPasteUtil.getFileList(t);
       if (fileList != null) {
+        boolean dropResult = ContainerUtil.process(Extensions.getExtensions(CustomFileDropHandler.CUSTOM_DROP_HANDLER_EP, project),
+                                                   handler -> !(handler.canHandle(t) &&
+                                                                handler.handleDrop(t, project)));
+        if (!dropResult) return;
+
         openFiles(project, fileList, editorWindow);
       }
     }
