Index: org/objectweb/asm/tree/MethodNode.java
===================================================================
--- org/objectweb/asm/tree/MethodNode.java
+++ org/objectweb/asm/tree/MethodNode.java
@@ -623,7 +623,7 @@
      */
     protected LabelNode getLabelNode(final Label l) {
         if (!(l.info instanceof LabelNode)) {
-            l.info = new LabelNode();
+            l.info = new LabelNode(l);  //[JB: passing a reference to avoid exceptions in coverage]
         }
         return (LabelNode) l.info;
     }
