Index: org/objectweb/asm/AnnotationVisitor.java
===================================================================
--- org/objectweb/asm/AnnotationVisitor.java
+++ org/objectweb/asm/AnnotationVisitor.java
@@ -76,7 +76,18 @@
         if (api < Opcodes.ASM4 || api > Opcodes.ASM6) {
             throw new IllegalArgumentException();
         }
-        this.api = api;
+        this.api = Opcodes.ASM6; //[JB: use the latest API version]
+        this.av = av;
+    }
+
+    //[JB: use the latest API version]
+    public AnnotationVisitor() {
+        this(null);
+    }
+
+    //[JB: use the latest API version]
+    public AnnotationVisitor(final AnnotationVisitor av) {
+        this.api = Opcodes.ASM6;
         this.av = av;
     }

Index: org/objectweb/asm/ClassVisitor.java
===================================================================
--- org/objectweb/asm/ClassVisitor.java
+++ org/objectweb/asm/ClassVisitor.java
@@ -78,7 +78,18 @@
         if (api < Opcodes.ASM4 || api > Opcodes.ASM6) {
             throw new IllegalArgumentException();
         }
-        this.api = api;
+        this.api = Opcodes.ASM6; //[JB: use the latest API version]
+        this.cv = cv;
+    }
+
+    //[JB: use the latest API version]
+    public ClassVisitor() {
+        this(null);
+    }
+
+    //[JB: use the latest API version]
+    public ClassVisitor(final ClassVisitor cv) {
+        this.api = Opcodes.ASM6;
         this.cv = cv;
     }
 
Index: org/objectweb/asm/FieldVisitor.java
===================================================================
--- org/objectweb/asm/FieldVisitor.java
+++ org/objectweb/asm/FieldVisitor.java
@@ -75,7 +75,18 @@
         if (api < Opcodes.ASM4 || api > Opcodes.ASM6) {
             throw new IllegalArgumentException();
         }
-        this.api = api;
+        this.api = Opcodes.ASM6; //[JB: use the latest API version]
+        this.fv = fv;
+    }
+
+    //[JB: use the latest API version]
+    public FieldVisitor() {
+        this(null);
+    }
+
+    //[JB: use the latest API version]
+    public FieldVisitor(final FieldVisitor fv) {
+        this.api = Opcodes.ASM6;
         this.fv = fv;
     }
 
Index: org/objectweb/asm/MethodVisitor.java
===================================================================
--- org/objectweb/asm/MethodVisitor.java
+++ org/objectweb/asm/MethodVisitor.java
@@ -92,7 +92,18 @@
         if (api < Opcodes.ASM4 || api > Opcodes.ASM6) {
             throw new IllegalArgumentException();
         }
-        this.api = api;
+        this.api = Opcodes.ASM6; //[JB: use the latest API version]
+        this.mv = mv;
+    }
+
+    //[JB: use the latest API version]
+    public MethodVisitor() {
+        this(null);
+    }
+
+    //[JB: use the latest API version]
+    public MethodVisitor(final MethodVisitor mv) {
+        this.api = Opcodes.ASM6;
         this.mv = mv;
     }
 
Index: org/objectweb/asm/ModuleVisitor.java
===================================================================
--- org/objectweb/asm/ModuleVisitor.java
+++ org/objectweb/asm/ModuleVisitor.java
@@ -68,7 +68,18 @@
         if (api != Opcodes.ASM6) {
             throw new IllegalArgumentException();
         }
-        this.api = api;
+        this.api = Opcodes.ASM6; //[JB: use the latest API version]
+        this.mv = mv;
+    }
+
+    //[JB: use the latest API version]
+    public ModuleVisitor() {
+        this(null);
+    }
+
+    //[JB: use the latest API version]
+    public ModuleVisitor(final ModuleVisitor mv) {
+        this.api = Opcodes.ASM6;
         this.mv = mv;
     }
     
Index: org/objectweb/asm/signature/SignatureVisitor.java
===================================================================
--- org/objectweb/asm/signature/SignatureVisitor.java
+++ org/objectweb/asm/signature/SignatureVisitor.java
@@ -88,7 +88,12 @@
         if (api < Opcodes.ASM4 || api > Opcodes.ASM6) {
             throw new IllegalArgumentException();
         }
-        this.api = api;
+        this.api = Opcodes.ASM6; //[JB: use the latest API version]
+    }
+
+    //[JB: use the latest API version]
+    public SignatureVisitor() {
+        this.api = Opcodes.ASM6;
     }
 
     /**
