Index: org/objectweb/asm/util/Textifier.java
===================================================================
--- org/objectweb/asm/util/Textifier.java
+++ org/objectweb/asm/util/Textifier.java
@@ -425,7 +425,7 @@
         }
 
         buf.append(tab);
-        appendAccess(access & ~Opcodes.ACC_VOLATILE);
+        appendAccess(access & ~(Opcodes.ACC_VOLATILE|Opcodes.ACC_TRANSIENT));  //[JB: IDEA-166121]
         if ((access & Opcodes.ACC_NATIVE) != 0) {
             buf.append("native ");
         }
