Index: community/jps/jps-builders/src/org/jetbrains/jps/dependency/java/Utils.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/community/jps/jps-builders/src/org/jetbrains/jps/dependency/java/Utils.java b/community/jps/jps-builders/src/org/jetbrains/jps/dependency/java/Utils.java
--- a/community/jps/jps-builders/src/org/jetbrains/jps/dependency/java/Utils.java	(revision 4cce80e49d99e4be40cbf6a7758e3629e78c39fe)
+++ b/community/jps/jps-builders/src/org/jetbrains/jps/dependency/java/Utils.java	(date 1742212993303)
@@ -60,7 +60,7 @@
     mySourcesFilter = sourceFilter;
     myIsNodeDeleted = isNodeDeleted;
     myDirectSubclasses = Objects.requireNonNull(myGraph.getIndex(SubclassesIndex.NAME));
-    myDeltaDirectSubclasses = myDelta != null? Objects.requireNonNull(myDelta.getIndex(SubclassesIndex.NAME)) : null;
+    myDeltaDirectSubclasses = myDelta == null || myDelta.isSourceOnly() ? null : Objects.requireNonNull(myDelta.getIndex(SubclassesIndex.NAME));
   }

   public Iterable<NodeSource> getNodeSources(ReferenceID nodeId) {
