--- /Users/denofevil/Code/sandbox/checker/syntax/relaxng/xhtml5full-xhtml-rdfa.rnc	2014-01-24 12:19:52.000000000 +0400
+++ html5/xhtml5full-xhtml-rdfa.rnc	2014-01-25 11:13:41.000000000 +0400
@@ -4,13 +4,17 @@
 		HTMLonly = notAllowed
 		v5only = empty
 }
+include "common-ext.rnc"
+
 start = html.elem
 include "phrase.rnc"
 include "block.rnc"
 include "sectional.rnc"
 include "revision.rnc"
 include "embed.rnc"
+include "embed-ext.rnc"
 include "core-scripting.rnc"
+include "core-scripting-ext.rnc"
 include "structural.rnc"
 include "ruby.rnc"
 include "media.rnc"
