--- /Users/denofevil/Code/sandbox/checker/validator/schema/html5-svg-mathml.rnc	2014-01-24 12:18:54.000000000 +0400
+++ html5-svg-mathml.rnc	2014-01-25 15:32:10.000000000 +0400
@@ -1,7 +1,7 @@
 namespace svg = "http://www.w3.org/2000/svg"
 namespace math = "http://www.w3.org/1998/Math/MathML"
 
-include "svg11/svg11-inc.rnc"
+include "svg11/svg11-tiny-inc.rnc"
 include "mml3/mathml3-inc.rnc"
 
 common.elem.phrasing |= svg
