Index: xml/relaxng/src/resources/html5-schema/html5/legacy.rnc
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- html5/legacy.rnc	(revision d1fb137d55cd5a5f1b9d6228d3bd49669e148540)
+++ html5/legacy.rnc	(revision aebff73f628a2e4a9321b0530831e1e36b280397)
@@ -142,6 +142,7 @@
 		element applet { applet.inner.phrasing & applet.attrs }
 	applet.attrs =
 		(	common.attrs
+		&	applet.attrs.alt?
 		&	applet.attrs.archive?
 		&	applet.attrs.code
 		&	applet.attrs.codebase?
@@ -149,6 +150,10 @@
 		&	applet.attrs.height
 		&	applet.attrs.width
 		)
+		applet.attrs.alt =
+			attribute alt {
+				string
+			}
 		applet.attrs.archive =
 			attribute archive {
 				common.data.uri #FIXME *comma* separated URI list (*grumble*)
