diff --git a/build/build.py b/build/build.py
index f3ab4a83..919a6576 100755
--- a/build/build.py
+++ b/build/build.py
@@ -1846,7 +1846,7 @@ def main(argv):
             elif arg == 'checkout':
                 pass
             elif arg == 'build':
-                release.buildAll()
+                buildSchemaDrivers()
             elif arg == 'docker-build':
                 dockerBuild()
             elif arg == 'docker-run':
