Index: community/xml/relaxng/src/resources/html5-schema/svg11/svg-gradient.rnc
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- svg11/svg-gradient.rnc	(revision 00d9e72224fc0433b68c80577d436341ef3f16fa)
+++ svg11/svg-gradient.rnc	(date 1563363151764)
@@ -62,7 +62,7 @@
         attribute x2 { Coordinate.datatype }?,
         attribute y2 { Coordinate.datatype }?,
         attribute gradientUnits {
-            string "userSpaceOnUse" | string "objectBoundingBox"
+            string "objectBoundingBox" | string "userSpaceOnUse"
         }?,
         attribute gradientTransform { TransformList.datatype }?,
         attribute spreadMethod { string "pad" | string "reflect" | string "repeat" }?
