From d48bebc211cc216aaa78bdf25d7f0b0143d6333b Mon Sep 17 00:00:00 2001
From: D D-D <aaaa@gmail.com>
Date: Wed, 12 Oct 2016 19:03:51 -0700
Subject: [PATCH 1/5] Subject line

---

mode change 100755 => 100644 l.php

diff --git a/l.php b/l.php
 old mode 100755
 new mode 100644
diff --git a/community/FileDropHandler.java b/community/FileDropHandler.java 
--- a/community/FileDropHandler.java
+++ b/community/FileDropHandler.java
@@ -1,9 +1,11 @@
 package com.intellij.openapi.fileEditor.impl.text;

 import com.intellij.ide.dnd.FileCopyPasteUtil;
+import com.intellij.openapi.editor.CustomFileDropHandler;
 import com.intellij.openapi.editor.Document;
 import com.intellij.openapi.editor.Editor;
 import com.intellij.openapi.editor.EditorDropHandler;
+import com.intellij.openapi.extensions.Extensions;
 import com.intellij.openapi.fileEditor.*;
 import com.intellij.openapi.fileEditor.ex.FileEditorManagerEx;
 import com.intellij.openapi.fileEditor.impl.EditorWindow;
